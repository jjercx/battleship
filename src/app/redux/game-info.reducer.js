import {
  GAME_SETUP,
  GAME_READY,
  GAME_END,
  SET_SIZE,
} from "app/constants/action-types";

export const getGameMode = ({ gameInfo: { gameMode } }) => gameMode;
export const getGameReady = ({ gameInfo: { ready } }) => ready;
export const getGameResult = ({ gameInfo: { result } }) => result;
export const getSize = ({ gameInfo: { size } }) => size;

const initialState = {
  gameMode: null,
  result: null,
  ready: false,
  numPlayers: 1,
  size: 10,
};

export default (state = initialState, { type, payload }) => {
  switch (type) {
    case SET_SIZE: {
      const { size } = payload;
      return { ...state, size };
    }
    case GAME_SETUP: {
      const { gameMode, numPlayers } = payload;
      return { ...state, gameMode, numPlayers, ready: false, result: null };
    }
    case GAME_READY: {
      return { ...state, ready: true };
    }
    case GAME_END: {
      const { result } = payload;
      return { ...state, result };
    }
    default:
      return state;
  }
};
