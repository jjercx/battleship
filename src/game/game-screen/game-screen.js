import React from "react";
import styled from "styled-components";
import Grid from "game/grid";
import Stat from "game/stat";
import shipData from "./data";
import Ship from "game/ship-item/ship";
import ShipItem from "game/ship-item/ship-item";
import Header from "app/header";

const Container = styled.div`
  padding-bottom: 20px;
  min-width: 100vw;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;
  background-color: ${props => props.theme.black};
`;

const StatPanel = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  margin: 10px;
`;

const ShipPanel = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-template-rows: repeat(5, 1fr);
  margin: 10px;
`;

const Row = styled.div`
  display: flex;
  flex-wrap: wrap-reverse;
  justify-content: center;
`;

const Column = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
`;

const ships = [];
shipData.forEach(({ count, size }) => {
  [...Array(count)].forEach((_, i) => {
    ships.push(new Ship(size));
  });
});

const GameScreen = ({ size = 10 }) => (
  <Container>
    <Header />
    <Row>
      <Column>
        <StatPanel>
          <Stat value={0} title="Hits" color="red"></Stat>
          <Stat value={0} title="Shots" color="green"></Stat>
          <Stat value={0} title="Turns" color="blue"></Stat>
        </StatPanel>
        <ShipPanel>
          {ships.map(ship => (
            <ShipItem ship={ship} key={ship.id}></ShipItem>
          ))}
        </ShipPanel>
      </Column>
      <Grid size={size} />
    </Row>
  </Container>
);

export default GameScreen;
