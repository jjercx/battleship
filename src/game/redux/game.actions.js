import {
  GAME_READY,
  GAME_SETUP,
  TILE_TOUCH,
  GAME_UPDATE,
  GAME_END,
  STORE_GAME_RECORD,
} from "app/constants/action-types";

export const gameSetup = ({ gameMode }) => ({
  type: GAME_SETUP,
  payload: gameMode,
});

export const gameReady = ({ ships, board }) => ({
  type: GAME_READY,
  payload: {
    ships,
    board,
  },
});

export const tileTouch = ({ row, col }) => ({
  type: TILE_TOUCH,
  payload: {
    row,
    col,
  },
});

export const gameUpdate = ({ ship, shots, hits, turns }) => ({
  type: GAME_UPDATE,
  payload: { ship, shots, hits, turns },
});

export const gameEnd = ({ result }) => ({
  type: GAME_END,
  payload: { result },
});

export const storeGameRecord = ({ gameRecord }) => ({
  type: STORE_GAME_RECORD,
  payload: { gameRecord },
});
