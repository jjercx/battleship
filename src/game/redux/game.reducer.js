import {
  GAME_SETUP,
  GAME_READY,
  GAME_UPDATE,
  GAME_END,
  SET_SIZE,
} from "app/constants/action-types";
import Ship from "game/models/ship";
import * as gameResult from "../constants/game-result";

export const getShips = ({ game: { ships } }) =>
  Object.values(ships).map(ship => new Ship(ship));

export const getShip = ({ game: { ships } }, shipId) => ships[shipId];

export const getHits = ({ game: { hits } }) => hits;

export const getShots = ({ game: { shots } }) => shots;

export const getTurns = ({ game: { turns } }) => turns;

export const getSize = ({ game: { size } }) => size;

export const getGameResult = ({ game: { result } }) => result;

export const getCell = ({ game: { board } }, row, col) =>
  board && board[row][col];

export const initialState = {
  board: null,
  ships: {},
  hits: 0,
  shots: 0,
  turns: 0,
  size: 10,
  result: null,
};

export default (state = initialState, { type, payload }) => {
  switch (type) {
    case SET_SIZE: {
      const { size } = payload;
      return { ...state, size };
    }
    case GAME_SETUP: {
      const gameMode = payload;
      const { turns } = gameMode;
      return { ...state, turns, shots: 0, hits: 0 };
    }
    case GAME_READY: {
      const { ships, board } = payload;
      return {
        ...state,
        board,
        ships,
      };
    }
    case GAME_UPDATE: {
      const { ship, shots, hits, turns } = payload;

      const newShips = ship
        ? { ...state.ships, [ship.id]: ship }
        : { ...state.ships };

      const newHits = hits ?? state.hits;

      return {
        ...state,
        ships: newShips,
        shots,
        hits: newHits,
        turns,
      };
    }
    case GAME_END: {
      const { win } = payload;
      const result = win ? gameResult.WIN : gameResult.LOSE;
      return { ...state, result };
    }
    default:
      return state;
  }
};
