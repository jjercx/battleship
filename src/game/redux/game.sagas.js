import { takeLatest, put, takeEvery, select } from "redux-saga/effects";
import {
  GAME_SETUP,
  TILE_TOUCH,
  GAME_UPDATE,
  GAME_END,
} from "app/constants/action-types";
import Game from "game/models/game";
import {
  gameReady,
  gameUpdate,
  gameEnd,
  storeGameRecord,
} from "game/redux/game.actions";
import {
  getCell,
  getShip,
  getShots,
  getHits,
  getTurns,
  getShips,
} from "./game.reducer";
import GameRecord from "leaderboard/models/game-record";
import * as gameResult from "game/constants/game-result";
import {
  getGameMode,
  getSize,
  getNumPlayers,
} from "app/redux/game-info.reducer";

export function* onGameSetup() {
  const boardSize = yield select(getSize);
  const numPlayers = yield select(getNumPlayers);
  console.log("numPlayers: ", numPlayers);
  const games = Game.setup(boardSize, numPlayers);
  yield put(gameReady({ games }));
}

export function* onTileTouch({ payload: { row, col, player } }) {
  const shipId = yield select(getCell, row, col, player);
  const ship = yield select(getShip, shipId, player);

  let shots = yield select(getShots, player);
  let hits = yield select(getHits, player);
  let turns = yield select(getTurns, player);

  shots += 1;
  turns -= 1;

  if (ship) {
    ship.hit();
    hits += 1;

    yield put(gameUpdate({ ship, shots, hits, turns, player }));
    return;
  }

  yield put(gameUpdate({ shots, turns, player }));
}

export function* watchForGameEnd() {
  const numPlayers = yield select(getNumPlayers);

  for (let player = 1; player === numPlayers; player++) {
    const ships = yield select(getShips, player);
    const allDead = ships.every(ship => !ship.isAlive());

    if (allDead) {
      if (numPlayers === 1) {
        yield put(gameEnd({ result: gameResult.WIN }));
        return;
      }

      yield put(gameEnd({ result: gameResult[`WIN_P${player}`] }));
      return;
    }

    const turns = yield select(getTurns, player);
    if (turns === 0) {
      if (numPlayers === 1) {
        yield put(gameEnd({ result: gameResult.LOSE }));
        return;
      }

      if (player === 1) {
        yield put(gameEnd({ result: gameResult.WIN_P2 }));
        return;
      }

      yield put(gameEnd({ result: gameResult.WIN_P1 }));
      return;
    }
  }
}

export function* onGameEnd({ payload: { result } }) {
  const numPlayers = yield select(getNumPlayers);

  if (numPlayers > 1) {
    return;
  }

  const shots = yield select(getShots, 1);
  const { turns, name: mode } = yield select(getGameMode);
  const gameRecord = new GameRecord({ shots, turns, mode, result });
  yield put(storeGameRecord({ gameRecord }));
}

export default [
  takeLatest(GAME_SETUP, onGameSetup),
  takeEvery(TILE_TOUCH, onTileTouch),
  takeEvery(GAME_UPDATE, watchForGameEnd),
  takeEvery(GAME_END, onGameEnd),
];
