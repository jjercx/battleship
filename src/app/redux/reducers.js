import { combineReducers } from "redux";
import game from "game/redux/game.reducer";

export default combineReducers({
  game,
});
