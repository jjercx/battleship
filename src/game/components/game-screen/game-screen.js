import React from "react";
import styled from "styled-components";
import Grid from "game/components/grid";
import Stat from "game/components/stat";
import ShipItem from "game/components/ship-item/ship-item";
import Modal from "app/components/modal";
import Button from "app/components/button";
import { connect } from "react-redux";
import {
  getShips,
  getLoading,
  getShots,
  getTurns,
  getHits,
  getSize,
} from "game/redux/game.reducer";
import * as actions from "game/redux/game.actions";
import { compose } from "redux";
import { withRouter } from "react-router-dom";
import * as routes from "app/constants/routes";

const Container = styled.div`
  position: relative;
  min-width: 100vw;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  background-color: ${props => props.theme.black};
`;

const StatPanel = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  margin: 10px;
`;

const ShipPanel = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-template-rows: repeat(5, 1fr);
  margin: 10px;
`;

const Row = styled.div`
  display: flex;
  flex-wrap: wrap-reverse;
  justify-content: center;
`;

const Column = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;

export const GameScreen = ({
  loading,
  ships = [],
  size,
  hits,
  shots,
  turns,
  history,
}) => {
  if (loading) {
    return null;
  }

  const handleExit = () => history.replace(routes.HOME);

  const handlePlayAgain = () => console.log("handlePlayAgain");

  const handleLeaderboard = () => console.log("handleLeaderboard");

  const handleChangeMode = () => console.log("handleChangeMode");

  return (
    <Container>
      <Row>
        <Column>
          <StatPanel>
            <Stat value={hits} title="Hits"></Stat>
            <Stat value={shots} title="Shots"></Stat>
            <Stat value={turns === Infinity ? "âˆž" : turns} title="Turns"></Stat>
          </StatPanel>
          <ShipPanel>
            {ships.map(ship => (
              <ShipItem ship={ship} key={ship.id}></ShipItem>
            ))}
          </ShipPanel>
        </Column>
        <Grid size={size} />
      </Row>
      <Modal title="game over">
        <Column>
          <Button text="play again" onClick={handlePlayAgain} />
          <Button text="change mode" onClick={handleChangeMode} />
          <Button text="leaderboard" onClick={handleLeaderboard} />
          <Button text="exit" onClick={handleExit} />
        </Column>
      </Modal>
    </Container>
  );
};

const mapStateToProps = state => ({
  loading: getLoading(state),
  ships: getShips(state),
  shots: getShots(state),
  turns: getTurns(state),
  hits: getHits(state),
  size: getSize(state),
});

export default compose(
  connect(mapStateToProps, actions),
  withRouter
)(GameScreen);
