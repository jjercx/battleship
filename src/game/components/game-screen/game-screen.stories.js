import React from "react";
import { GameScreen } from "./game-screen";

export default {
  title: "game/GameScreen",
  component: GameScreen,
};

export const Default = () => <GameScreen />;
