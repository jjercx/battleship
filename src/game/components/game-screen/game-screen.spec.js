import React from "react";
import { GameScreen } from "./game-screen";
import renderer from "react-test-renderer";
import { createStore } from "redux";
import rootReducer from "app/redux/reducers";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";

describe("<GameScreen />", () => {
  describe("with initialState", () => {
    it("matches snapshot", () => {
      const store = createStore(rootReducer);
      const tree = renderer
        .create(
          <Provider store={store}>
            <MemoryRouter>
              <GameScreen />
            </MemoryRouter>
          </Provider>
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
