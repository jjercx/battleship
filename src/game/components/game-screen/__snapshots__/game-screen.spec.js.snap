// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameScreen /> with initialState matches snapshot 1`] = `
<div
  className="sc-fznZeY gzVttC"
>
  <div
    className="sc-fzqNJr kxxivL"
  >
    <div
      className="sc-fzoyAV lgOAXd"
    >
      <div
        className="sc-fzokOt cLlFZR"
      >
        <div
          className="sc-AxgMl hlfqrj"
        >
          <span
            className="sc-AxheI emrIIm"
          />
          <span
            className="sc-Axmtr jfKBn"
          >
            Hits
          </span>
        </div>
        <div
          className="sc-AxgMl hlfqrj"
        >
          <span
            className="sc-AxheI emrIIm"
          />
          <span
            className="sc-Axmtr jfKBn"
          >
            Shots
          </span>
        </div>
        <div
          className="sc-AxgMl hlfqrj"
        >
          <span
            className="sc-AxheI emrIIm"
          />
          <span
            className="sc-Axmtr jfKBn"
          >
            Turns
          </span>
        </div>
      </div>
      <div
        className="sc-fzqBZW gcZzYR"
      />
    </div>
  </div>
</div>
`;
